transaction/std/src/memo_builder/gift_code_memo_builder.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Defines the Gift Code Memo Builder for gift code memos specified in MCIP #32
+
+use super::{memo::UnusedMemo, MemoBuilder};
+use crate::{
+    memo::{GiftCodeCancellationMemo, GiftCodeFundingMemo, GiftCodeSenderMemo},
+    ChangeDestination,
+};
+use mc_account_keys::PublicAddress;
+use mc_crypto_keys::RistrettoPublic;
+use mc_transaction_core::{Amount, MemoContext, MemoError, MemoPayload, NewMemoError};
+use std::convert::TryFrom;
+
+/// The three possible memo types that can be written as specified in MCIP #32
+/// | Memo type bytes | Name                                              |
+/// | -----------     | -----------                                       |
+/// | 0x0002          | Gift Code Sender Memo                             |
+/// | 0x0201          | Gift Code Funding Memo                            |
+/// | 0x0202          | Gift Code Cancellation Memo                       |
+#[derive(Clone, Debug)]
+pub enum GiftCodeMemoType {
+    /// Gift Code Sender Memo 0x0002
+    Sender,
+    /// Gift Code Funding Memo 0x0201
+    Funding,
+    /// Gift Code Cancellation Memo 0x0202
+    Cancellation,
+}
+
+/// This memo builder writes gift code memos to the change address which can be
+/// used by apps that offer gift codes as feature to track the status of gift
+/// codes as well as certain contextual information about them. This builder
+/// writes 3 possible gift code memo types. Only one memo which can be specified
+/// at the creation of the memo builder. It is worth noting that all memos
+/// written by this memo builder will go to the reserved change address and
+/// does not write any destination memos. For normal recoverable transaction
+/// history (RTH), the RTH memo builder should be used in conjuction with this
+/// memo builder.
+#[derive(Clone, Debug)]
+pub struct GiftCodeMemoBuilder {
+    // The gift code memo type we'll be writing
+    gift_code_memo_type: GiftCodeMemoType,
+    // The utf-8 formatted note that appears on the change address when a gift code is
+    // or cancelled
+    note: &'static str,
```

Comment:
What if the note comes at runtime, e.g. from the user inputting something interactively?

Reply:
A `String` is not `'static`:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b44493d157b3220be01831330cbe83ea